src/providers.rs

```diff
@@ -1,406 +1,259 @@
-pub const BLOCKPI_ARBITRUM_ONE_HOSTNAME: &str = "arbitrum.blockpi.network";
-pub const PUBLICNODE_ARBITRUM_ONE_HOSTNAME: &str = "arbitrum-one-rpc.publicnode.com";
-pub const ALCHEMY_BASE_MAINNET_HOSTNAME: &str = "base-mainnet.g.alchemy.com";
-pub const BLOCKPI_BASE_MAINNET_HOSTNAME: &str = "base.blockpi.network";
-pub const PUBLICNODE_BASE_MAINNET_HOSTNAME: &str = "base-rpc.publicnode.com";
-pub const ALCHEMY_OPT_MAINNET_HOSTNAME: &str = "opt-mainnet.g.alchemy.com";
-pub const BLOCKPI_OPTIMISM_MAINNET_HOSTNAME: &str = "optimism.blockpi.network";
-pub const PUBLICNODE_OPTIMISM_MAINNET_HOSTNAME: &str = "optimism-rpc.publicnode.com";
-pub const LLAMA_ETH_MAINNET_HOSTNAME: &str = "eth.llamarpc.com";
-pub const LLAMA_ARBITRUM_ONE_HOSTNAME: &str = "arbitrum.llamarpc.com";
-pub const LLAMA_BASE_MAINNET_HOSTNAME: &str = "base.llamarpc.com";
-pub const LLAMA_OPTIMISM_MAINNET_HOSTNAME: &str = "optimism.llamarpc.com";
+const ALCHEMY_HEADERS: &[ConstHeader] = &[ConstHeader {
+    name: "Authorization",
+    value: "Bearer {API_KEY}",
+}];
 
-// Limited API credentials for local testing.
-// Use `dfx canister call evm_rpc updateProvider ...` to pass your own keys.
-pub const ALCHEMY_ETH_MAINNET_CREDENTIAL: &str = "/v2/zBxaSBUMfuH8XnA-uLIWeXfCx1T8ItkM";
-pub const ALCHEMY_ETH_SEPOLIA_CREDENTIAL: &str = "/v2/Mbow19DWsfPXiTpdgvRu4HQq63iYycU-";
-pub const ALCHEMY_ARBITRUM_ONE_CREDENTIAL: &str = "/v2";
-pub const ALCHEMY_BASE_MAINNET_CREDENTIAL: &str = "/v2";
-pub const ALCHEMY_OPTIMISM_MAINNET_CREDENTIAL: &str = "/v2";
-pub const BLOCKPI_ETH_MAINNET_CREDENTIAL: &str = "/v1/rpc/0edc81e20be23ddff051f61a97bb457ec7284a58";
-pub const BLOCKPI_ETH_SEPOLIA_CREDENTIAL: &str = "/v1/rpc/1fe987fddded17db50862311720ff444991d4dab";
-pub const BLOCKPI_ARBITRUM_ONE_CREDENTIAL: &str =
-    "/v1/rpc/a8b89a41d2a341e32ee7aefcb20820a7cbb65f35";
-pub const BLOCKPI_BASE_MAINNET_CREDENTIAL: &str =
-    "/v1/rpc/bd458bf9f28ed45c77823814a937c812d2efd260";
-pub const BLOCKPI_OPTIMISM_MAINNET_CREDENTIAL: &str =
-    "/v1/rpc/d54bfe59299d56b0cbb8b3c69bd122f4ab5ac654";
+pub const PROVIDERS: &[Provider] = &[
+    Provider {
+        provider_id: 0,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        public_url: Some("https://cloudflare-eth.com/v1/mainnet"),
+        url_pattern: "https://cloudflare-eth.com/v1/mainnet/{API_KEY}",
+        header_patterns: &[],
+        service: Some(RpcService::EthMainnet(EthMainnetService::Cloudflare)),
+    },
+    Provider {
+        provider_id: 1,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        public_url: None,
+        url_pattern: "https://rpc.ankr.com/eth/{API_KEY}",
+        header_patterns: &[],
+        service: Some(RpcService::EthMainnet(EthMainnetService::Ankr)),
+    },
+    Provider {
+        provider_id: 2,
+        chain_id: ETH_MAINNET_CHAIN_ID,
+        public_url: Some("https://ethereum-rpc.publicnode.com"),
+        url_pattern: "https://ethereum-rpc.publicnode.com",
+        header_patterns: &[],
```

Comment:
Good question! The new `Provider` has basically 3 moving parts (`public_url`, `url_pattern`, `header_patterns`) related to how the provider is accessed so I suggest to model this more explicitly. Since Authentication via header is always done by Bearer Token, I suggest the following approach (also ok to do in another PR if you think that's simpler). Please feel free to comment/challenge

```rust
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Provider2 {
    pub provider_id: ProviderId,
    pub chain_id: u64,
    pub access: RpcServiceAccess,
    pub alias: Option<RpcService>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RpcServiceAccess {
    Authenticated {
        auth_type: RpcServiceAuth,
        /// Public URL to use when the API key is not available.
        fallback_public_url: Option<&'static str>,
    },
    Unauthenticated {
        public_url: &'static str,
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RpcServiceAuth {
    /// API Key will be used in an Authorization header as Bearer token, e.g.,
    /// `Authorization: Bearer API_KEY`
    BearerToken,
    UrlParameter { url_pattern: &'static str },
}
```

With this the `PROVIDERS` would look like this (just took 3 as an example of the various use-cases)
```rust
pub const PROVIDERS2: &[Provider2] = &[
    Provider2 {
        provider_id: 3,
        chain_id: ETH_MAINNET_CHAIN_ID,
        access: RpcServiceAccess::Authenticated {
            auth_type: RpcServiceAuth::UrlParameter {
                url_pattern: "https://ethereum.blockpi.network/v1/rpc/{API_KEY}",
            },
            fallback_public_url: Some("https://ethereum.blockpi.network/v1/rpc/public"),
        },
        alias: Some(RpcService::EthMainnet(EthMainnetService::BlockPi)),
    },
    Provider2 {
        provider_id: 8,
        chain_id: ETH_MAINNET_CHAIN_ID,
        access: RpcServiceAccess::Authenticated {
            auth_type: RpcServiceAuth::BearerToken,
            fallback_public_url: None,
        },
        alias: Some(RpcService::EthMainnet(EthMainnetService::Alchemy)),
    },
    Provider2 {
        provider_id: 22,
        chain_id: ETH_MAINNET_CHAIN_ID,
        access: RpcServiceAccess::Unauthenticated {
            public_url: "https://eth.llamarpc.com",
        },
        alias: Some(RpcService::EthMainnet(EthMainnetService::Llama)),
    },
];
```

WDYT?

Reply:
I like this idea in principle. There are some edge cases with the proposed design, e.g. using `RpcServiceAccess::Authenticated` with `RpcServiceAuth::BearerToken` has no way to specify the endpoint URL. 

How about the following modified types (with some subjective naming changes):

```rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Provider2 {
    pub provider_id: ProviderId,
    pub chain_id: u64,
    pub access: RpcAccess,
    pub alias: Option<RpcService>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RpcAccess {
    Authenticated {
        auth: RpcAuth,
        /// Public URL to use when the API key is not available.
        public_url: Option<&'static str>,
    },
    Unauthenticated {
        public_url: &'static str,
    },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum RpcAuth {
    /// API Key will be used in an Authorization header as Bearer token, e.g.,
    /// `Authorization: Bearer API_KEY`
    BearerToken { url: &'static str },
    UrlParameter { url_pattern: &'static str },
}
```

Reply:
> e.g. using RpcServiceAccess::Authenticated with RpcServiceAuth::BearerToken has no way to specify the endpoint URL.

Oups, that's something I forgot, good catch! Your suggestion LGTM!

Reply:
Updated to use this design. 

I also managed to further simplify the type definitions by removing `ProviderView` and deriving `Serialize` (but not `Deserialize`) for the `Provider`, `RpcAccess`, and `RpcAuth` structs. This makes me a lot happier with the type definitions; significantly less duplicated code.